File: dawn/src/main/java/com/lzq/dawn/util/phone/PhoneUtils.java
package com.lzq.dawn.util.phone;

import static android.Manifest.permission.CALL_PHONE;
import static android.Manifest.permission.READ_PHONE_STATE;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Build;
import android.telephony.TelephonyManager;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresPermission;

import com.lzq.dawn.DawnBridge;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * @Name :PhoneUtils
 * @Time :2022/8/16 14:29
 * @Author :  Lzq
 * @Desc : 手机
 */
public final class PhoneUtils {
    private PhoneUtils() {
    }

    /**
     * Return 设备是否为手机。
     *
     * @return {@code true}: yes<br>{@code false}: no
     */
    public static boolean isPhone() {
        TelephonyManager tm = getTelephonyManager();
        return tm.getPhoneType() != TelephonyManager.PHONE_TYPE_NONE;
    }

    /**
     * Return 唯一的设备 ID。
     * <p>如果SDK版本大于28，则返回空字符串。<p>
     * <p>必须持有 {@code <uses-permission android:name="android.permission.READ_PHONE_STATE" />}</p>
     *
     * @return 唯一的设备 ID。
     */
    @SuppressLint("HardwareIds")
    @RequiresPermission(READ_PHONE_STATE)
    public static String getDeviceId() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return "";
        }
        TelephonyManager tm = getTelephonyManager();
        String deviceId = tm.getDeviceId();
        if (!TextUtils.isEmpty(deviceId)) {
            return deviceId;
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String imei = tm.getImei();
            if (!TextUtils.isEmpty(imei)) {
                return imei;
            }
            String meid = tm.getMeid();
            return TextUtils.isEmpty(meid) ? "" : meid;
        }
        return "";
    }

    /**
     * Return 设备的序列号。
     *
     * @return 设备的序列号。
     */
    @SuppressLint("HardwareIds")
    @RequiresPermission(READ_PHONE_STATE)
    public static String getSerial() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            try {
                return Build.getSerial();
            } catch (SecurityException e) {
                e.printStackTrace();
                return "";
            }
        }
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ? Build.getSerial() : Build.SERIAL;
    }

    /**
     * Return IMEI
     * <p>如果SDK版本大于28，则返回空字符串。<p>
     * <p>必须持有{@code <uses-permission android:name="android.permission.READ_PHONE_STATE" />}</p>
     *
     * @return IMEI
     */
    @RequiresPermission(READ_PHONE_STATE)
    public static String getIMEI() {
        return getImeiOrMeid(true);
    }

    /**
     * Return  MEID.
     * <p>如果SDK版本大于28，则返回空字符串。</p>
     * <p>必须持有 {@code <uses-permission android:name="android.permission.READ_PHONE_STATE" />}</p>
     *
     * @return MEID
     */
    @RequiresPermission(READ_PHONE_STATE)
    public static String getMEID() {
        return getImeiOrMeid(false);
    }

    @SuppressLint("HardwareIds")
    @RequiresPermission(READ_PHONE_STATE)
    public static String getImeiOrMeid(boolean isImei) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return "";
        }
        TelephonyManager tm = getTelephonyManager();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (isImei) {
                return getMinOne(tm.getImei(0), tm.getImei(1));
            } else {
                return getMinOne(tm.getMeid(0), tm.getMeid(1));
            }
        } else {
            String ids = getSystemPropertyByReflect(isImei ? "ril.gsm.imei" : "ril.cdma.meid");
            if (!TextUtils.isEmpty(ids)) {
                String[] idArr = ids.split(",");
                if (idArr.length == 2) {
                    return getMinOne(idArr[0], idArr[1]);
                } else {
                    return idArr[0];
                }
            }

            String id0 = tm.getDeviceId();
            String id1 = "";
            try {
                Method method = tm.getClass().getMethod("getDeviceId", int.class);
                id1 = (String) method.invoke(tm,
                        isImei ? TelephonyManager.PHONE_TYPE_GSM
                                : TelephonyManager.PHONE_TYPE_CDMA);
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
            if (isImei) {
                if (id0 != null && id0.length() < 15) {
                    id0 = "";
                }
                if (id1 != null && id1.length() < 15) {
                    id1 = "";
                }
            } else {
                if (id0 != null && id0.length() == 14) {
                    id0 = "";
                }
                if (id1 != null && id1.length() == 14) {
                    id1 = "";
                }
            }
            return getMinOne(id0, id1);
        }
    }

    private static String getMinOne(String s0, String s1) {
        boolean empty0 = TextUtils.isEmpty(s0);
        boolean empty1 = TextUtils.isEmpty(s1);
        if (empty0 && empty1) {
            return "";
        }
        if (!empty0 && !empty1) {
            if (s0.compareTo(s1) <= 0) {
                return s0;
            } else {
                return s1;
            }
        }
        if (!empty0) {
            return s0;
        }
        return s1;
    }

    private static String getSystemPropertyByReflect(String key) {
        try {
            @SuppressLint("PrivateApi")
            Class<?> clz = Class.forName("android.os.SystemProperties");
            Method getMethod = clz.getMethod("get", String.class, String.class);
            return (String) getMethod.invoke(clz, key, "");
        } catch (Exception e) {/**/}
        return "";
    }

    /**
     * Return IMSI.
     * <p>必须持有 {@code <uses-permission android:name="android.permission.READ_PHONE_STATE" />}</p>
     *
     * @return the IMSI
     */
    @SuppressLint("HardwareIds")
    @RequiresPermission(READ_PHONE_STATE)
    public static String getIMSI() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            try {
                getTelephonyManager().getSubscriberId();
            } catch (SecurityException e) {
                e.printStackTrace();
                return "";
            }
        }
        return getTelephonyManager().getSubscriberId();
    }

    /**
     * Returns 当前电话类型。
     *
     * @return the current phone type
     * <ul>
     * <li>{@link TelephonyManager#PHONE_TYPE_NONE}</li>
     * <li>{@link TelephonyManager#PHONE_TYPE_GSM }</li>
     * <li>{@link TelephonyManager#PHONE_TYPE_CDMA}</li>
     * <li>{@link TelephonyManager#PHONE_TYPE_SIP }</li>
     * </ul>
     */
    public static int getPhoneType() {
        TelephonyManager tm = getTelephonyManager();
        return tm.getPhoneType();
    }

    /**
     * Return sim卡状态是否就绪。
     *
     * @return {@code true}: yes<br>{@code false}: no
     */
    public static boolean isSimCardReady() {
        TelephonyManager tm = getTelephonyManager();
        return tm.getSimState() == TelephonyManager.SIM_STATE_READY;
    }

    /**
     * Return sim 操作员名称。
     *
     * @return sim 操作员名称。
     */
    public static String getSimOperatorName() {
        TelephonyManager tm = getTelephonyManager();
        return tm.getSimOperatorName();
    }

    /**
     * Return mnc 返回 sim 运算符。
     *
     * @return sim 运算符。
     */
    public static String getSimOperatorByMnc() {
        TelephonyManager tm = getTelephonyManager();
        String operator = tm.getSimOperator();
        if (operator == null) {
            return "";
        }
        switch (operator) {
            case "46000":
            case "46002":
            case "46007":
            case "46020":
                return "中国移动";
            case "46001":
            case "46006":
            case "46009":
                return "中国联通";
            case "46003":
            case "46005":
            case "46011":
                return "中国电信";
            default:
                return operator;
        }
    }

    /**
     * 跳到拨号。
     *
     * @param phoneNumber 电话号码。
     */
    public static void dial(@NonNull final String phoneNumber) {
        DawnBridge.getApp().startActivity(DawnBridge.getDialIntent(phoneNumber));
    }

    /**
     * 直接拨号
     * <p>必须持有{@code <uses-permission android:name="android.permission.CALL_PHONE" />}</p>
     *
     * @param phoneNumber 电话号码。
     */
    @RequiresPermission(CALL_PHONE)
    public static void call(@NonNull final String phoneNumber) {
        DawnBridge.getApp().startActivity(DawnBridge.getCallIntent(phoneNumber));
    }

    /**
     * 发短信
     *
     * @param phoneNumber 电话号码
     * @param content     content.
     */
    public static void sendSms(@NonNull final String phoneNumber, final String content) {
        DawnBridge.getApp().startActivity(DawnBridge.getSendSmsIntent(phoneNumber, content));
    }

    private static TelephonyManager getTelephonyManager() {
        return (TelephonyManager) DawnBridge.getApp().getSystemService(Context.TELEPHONY_SERVICE);
    }
}

File: dawn/src/main/java/com/lzq/dawn/util/phone/PhoneUtils.kt
package com.lzq.dawn.util.phone

import android.Manifest.permission
import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.telephony.TelephonyManager
import android.text.TextUtils
import androidx.annotation.RequiresPermission
import com.lzq.dawn.DawnBridge
import java.lang.reflect.InvocationTargetException

/**
 * @Name :PhoneUtils
 * @Time :2022/8/16 14:29
 * @Author :  Lzq
 * @Desc : 手机
 */
object PhoneUtils {
    val isPhone: Boolean
        /**
         * Return 设备是否为手机。
         *
         * @return `true`: yes<br></br>`false`: no
         */
        get() {
            val tm = telephonyManager
            return tm.phoneType != TelephonyManager.PHONE_TYPE_NONE
        }

    @get:RequiresPermission(permission.READ_PHONE_STATE)
    @get:SuppressLint("HardwareIds")
    val deviceId: String
        /**
         * Return 唯一的设备 ID。
         *
         * 如果SDK版本大于28，则返回空字符串。
         *
         *
         *
         * 必须持有 `<uses-permission android:name="android.permission.READ_PHONE_STATE" />`
         *
         * @return 唯一的设备 ID。
         */
        get() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                return ""
            }
            val tm = telephonyManager
            val deviceId = tm.deviceId
            if (!TextUtils.isEmpty(deviceId)) {
                return deviceId
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val imei = tm.imei
                if (!TextUtils.isEmpty(imei)) {
                    return imei
                }
                val meid = tm.meid
                return if (TextUtils.isEmpty(meid)) "" else meid
            }
            return ""
        }

    @get:RequiresPermission(permission.READ_PHONE_STATE)
    @get:SuppressLint("HardwareIds")
    val serial: String
        /**
         * Return 设备的序列号。
         *
         * @return 设备的序列号。
         */
        get() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                return try {
                    Build.getSerial()
                } catch (e: SecurityException) {
                    e.printStackTrace()
                    ""
                }
            }
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) Build.getSerial() else Build.SERIAL
        }

    @get:RequiresPermission(permission.READ_PHONE_STATE)
    val iMEI: String?
        /**
         * Return IMEI
         *
         * 如果SDK版本大于28，则返回空字符串。
         *
         *
         *
         * 必须持有`<uses-permission android:name="android.permission.READ_PHONE_STATE" />`
         *
         * @return IMEI
         */
        get() = getImeiOrMeid(true)

    @get:RequiresPermission(permission.READ_PHONE_STATE)
    val mEID: String?
        /**
         * Return  MEID.
         *
         * 如果SDK版本大于28，则返回空字符串。
         *
         * 必须持有 `<uses-permission android:name="android.permission.READ_PHONE_STATE" />`
         *
         * @return MEID
         */
        get() = getImeiOrMeid(false)

    @SuppressLint("HardwareIds")
    @RequiresPermission(permission.READ_PHONE_STATE)
    fun getImeiOrMeid(isImei: Boolean): String? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return ""
        }
        val tm = telephonyManager
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (isImei) {
                getMinOne(tm.getImei(0), tm.getImei(1))
            } else {
                getMinOne(tm.getMeid(0), tm.getMeid(1))
            }
        } else {
            val ids = getSystemPropertyByReflect(if (isImei) "ril.gsm.imei" else "ril.cdma.meid")
            if (!TextUtils.isEmpty(ids)) {
                val idArr = ids.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                return if (idArr.size == 2) {
                    getMinOne(idArr[0], idArr[1])
                } else {
                    idArr[0]
                }
            }
            var id0 = tm.deviceId
            var id1 = ""
            try {
                val method = tm.javaClass.getMethod("getDeviceId", Int::class.javaPrimitiveType)
                id1 = method.invoke(
                    tm, if (isImei) TelephonyManager.PHONE_TYPE_GSM else TelephonyManager.PHONE_TYPE_CDMA
                ) as String
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            }
            if (isImei) {
                if (id0 != null && id0.length < 15) {
                    id0 = ""
                }
                if (id1 != null && id1.length < 15) {
                    id1 = ""
                }
            } else {
                if (id0 != null && id0.length == 14) {
                    id0 = ""
                }
                if (id1 != null && id1.length == 14) {
                    id1 = ""
                }
            }
            getMinOne(id0, id1)
        }
    }

    private fun getMinOne(s0: String?, s1: String?): String? {
        val empty0 = TextUtils.isEmpty(s0)
        val empty1 = TextUtils.isEmpty(s1)
        if (empty0 && empty1) {
            return ""
        }
        if (!empty0 && !empty1) {
            return if (s0!!.compareTo(s1!!) <= 0) {
                s0
            } else {
                s1
            }
        }
        return if (!empty0) {
            s0
        } else s1
    }

    private fun getSystemPropertyByReflect(key: String): String {
        try {
            @SuppressLint("PrivateApi") val clz = Class.forName("android.os.SystemProperties")
            val getMethod = clz.getMethod("get", String::class.java, String::class.java)
            return getMethod.invoke(clz, key, "") as String
        } catch (e: Exception) { /**/
        }
        return ""
    }

    @get:RequiresPermission(permission.READ_PHONE_STATE)
    @get:SuppressLint("HardwareIds")
    val iMSI: String
        /**
         * Return IMSI.
         *
         * 必须持有 `<uses-permission android:name="android.permission.READ_PHONE_STATE" />`
         *
         * @return the IMSI
         */
        get() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                try {
                    telephonyManager.subscriberId
                } catch (e: SecurityException) {
                    e.printStackTrace()
                    return ""
                }
            }
            return telephonyManager.subscriberId
        }
    val phoneType: Int
        /**
         * Returns 当前电话类型。
         *
         * @return the current phone type
         *
         *  * [TelephonyManager.PHONE_TYPE_NONE]
         *  * [TelephonyManager.PHONE_TYPE_GSM]
         *  * [TelephonyManager.PHONE_TYPE_CDMA]
         *  * [TelephonyManager.PHONE_TYPE_SIP]
         *
         */
        get() {
            val tm = telephonyManager
            return tm.phoneType
        }
    val isSimCardReady: Boolean
        /**
         * Return sim卡状态是否就绪。
         *
         * @return `true`: yes<br></br>`false`: no
         */
        get() {
            val tm = telephonyManager
            return tm.simState == TelephonyManager.SIM_STATE_READY
        }
    val simOperatorName: String
        /**
         * Return sim 操作员名称。
         *
         * @return sim 操作员名称。
         */
        get() {
            val tm = telephonyManager
            return tm.simOperatorName
        }
    val simOperatorByMnc: String
        /**
         * Return mnc 返回 sim 运算符。
         *
         * @return sim 运算符。
         */
        get() {
            val tm = telephonyManager
            val operator = tm.simOperator ?: return ""
            return when (operator) {
                "46000", "46002", "46007", "46020" -> "中国移动"
                "46001", "46006", "46009" -> "中国联通"
                "46003", "46005", "46011" -> "中国电信"
                else -> operator
            }
        }

    /**
     * 跳到拨号。
     *
     * @param phoneNumber 电话号码。
     */
    fun dial(phoneNumber: String) {
        DawnBridge.getApp().startActivity(DawnBridge.getDialIntent(phoneNumber))
    }

    /**
     * 直接拨号
     *
     * 必须持有`<uses-permission android:name="android.permission.CALL_PHONE" />`
     *
     * @param phoneNumber 电话号码。
     */
    @RequiresPermission(permission.CALL_PHONE)
    fun call(phoneNumber: String) {
        DawnBridge.getApp().startActivity(DawnBridge.getCallIntent(phoneNumber))
    }

    /**
     * 发短信
     *
     * @param phoneNumber 电话号码
     * @param content     content.
     */
    fun sendSms(phoneNumber: String, content: String?) {
        DawnBridge.getApp().startActivity(DawnBridge.getSendSmsIntent(phoneNumber, content))
    }

    private val telephonyManager: TelephonyManager
        private get() = DawnBridge.getApp().getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
}